consult-xref.el

```diff
@@ -0,0 +1,103 @@
+
+(defun xref-show-definitions-consult (fetcher alist)
+  "Let the user choose the target definition with completion.
+
+When there is more than one definition, let the user choose
+between them by typing in the minibuffer with completion."
+  (let* ((xrefs (funcall fetcher))
+         (xref-alist (xref--analyze xrefs))
+         xref-alist-with-line-info
+         xref
+         (group-prefix-length
+          ;; FIXME: Groups are not always file names, but they often
+          ;; are.  At least this shouldn't make the other kinds of
+          ;; groups look worse.
+          (let ((common-prefix (try-completion "" xref-alist)))
+            (if (> (length common-prefix) 0)
+                (length (file-name-directory common-prefix))
+              0))))
+
+    (cl-loop for ((group . xrefs) . more1) on xref-alist
+             do
+             (cl-loop for (xref . more2) on xrefs do
+                      (with-slots (summary location) xref
+                        (let* ((line (xref-location-line location))
+                               (line-fmt
+                                (if line
+                                    (format #("%d:" 0 2 (face xref-line-number))
+                                            line)
+                                  ""))
+                               (group-fmt
+                                (propertize
+                                 (substring group group-prefix-length)
+                                 'face 'xref-file-header))
+                               (candidate
+                                (format "%s:%s%s" group-fmt line-fmt summary)))
+                          (push (cons candidate xref) xref-alist-with-line-info)))))
+
+    (setq xref (if (not (cdr xrefs))
+                   (car xrefs)
+                 (let* ((collection (reverse xref-alist-with-line-info))
+                        (def (caar collection)))
+                   (cdr (assoc (consult--read collection
+                                              :prompt "xref results: "
+                                              :history nil
+                                              :require-match t
+                                              :predicate nil
+                                              :default def
+                                              :state (consult--xref-preview collection))
+                               collection)))))
+    (xref-pop-to-location xref (assoc-default 'display-action alist))))
+
+(defun consult--xref-preview (collection)
+  "Preview the xref result from consult.
+
+COLLECTION is the collection of definition candidates given by xref."
+  (let ((preview (consult--jump-preview))
+        (open (consult--temporary-files)))
+    (lambda (cand restore)
+      (cond
+       (restore
+        (funcall open)
+        (funcall preview nil t))
+       (t (let ((cand-xref (cdr (assoc cand collection))))
+                   (xref-pop-to-location cand-xref nil)))))))
```

Comment:
Instead of pop-to-location, preview and open should be used for a robust preview.